domain/src/main/java/org/oppia/android/domain/exploration/ExplorationActiveTimeController.kt

```diff
@@ -0,0 +1,182 @@
+
+  private var startExplorationTimestampMs: Long = 0L
+
+  /**
+   * Begin tracking the active learning time in an exploration.
+   *
+   * We define the active time loosely as the time spent in an exploration when the app is in the
+   * foreground.
+   *
+   * This method is called when the [ExplorationProgressController.beginExplorationImpl] finishes
+   * executing successfully, or when the app comes back to the foreground after being previously
+   * backgrounded.
+   */
+  fun setExplorationSessionStarted() {
+    this.startExplorationTimestampMs = oppiaClock.getCurrentTimeMs()
+  }
+
+  /**
+   * Stops tracking the active learning time in an exploration.
+   *
+   * This method is called when the [ExplorationProgressController.finishExplorationImpl] finishes
+   * executing, or the app goes to the background.
+   */
+  fun setExplorationSessionStopped(profileId: ProfileId, topicId: String) {
+    recordAggregateTopicLearningTime(
+      profileId = profileId,
+      topicId = topicId,
+      sessionDuration = oppiaClock.getCurrentTimeMs() - startExplorationTimestampMs
+    )
+  }
+
+  /**
+   * Records the tracked active exploration time.
+   *
+   * @param profileId the ID corresponding to the profile for which progress needs to be stored
+   * @param topicId the ID corresponding to the topic for which duration needs to be stored
+   * @param sessionDuration the tracked exploration duration between start and pause
+   * @return a [DataProvider] that indicates the success/failure of this record operation.
+   */
+  private fun recordAggregateTopicLearningTime(
+    profileId: ProfileId,
+    topicId: String,
+    sessionDuration: Long
+  ): DataProvider<Any?> {
+    val deferred = retrieveCacheStore(profileId).storeDataWithCustomChannelAsync(
+      updateInMemoryCache = true
+    ) { topicLearningTimeDatabase ->
+
+      val previousAggregateLearningTime =
+        topicLearningTimeDatabase.aggregateTopicLearningTimeMap[topicId]
+
+      val topicLearningTimeBuilder = if (previousAggregateLearningTime != null) {
+        previousAggregateLearningTime.toBuilder()
+      } else {
+        TopicLearningTime.newBuilder()
+          .setTopicId(topicId)
+          .setLastUpdatedTimeMs(oppiaClock.getCurrentTimeMs())
+      }
+
+      if (isLastUpdatedTimestampStale(topicLearningTimeBuilder.lastUpdatedTimeMs)
+      ) {
+        topicLearningTimeBuilder.topicLearningTimeMs = sessionDuration
+      } else {
+        topicLearningTimeBuilder.topicLearningTimeMs += sessionDuration
+      }
+      topicLearningTimeBuilder.lastUpdatedTimeMs = oppiaClock.getCurrentTimeMs()
+
+      val topicLearningTime = topicLearningTimeBuilder.build()
+
+      val topicLearningTimeDatabaseBuilder = topicLearningTimeDatabase.toBuilder()
+        .putAggregateTopicLearningTime(topicId, topicLearningTime)
+      Pair(topicLearningTimeDatabaseBuilder.build(), TopicLearningTimeActionStatus.SUCCESS)
```

Comment:
```suggestion
      topicLearningTimeDatabase.toBuilder().apply {
        val topicLearningTime = aggregateTopicLearningTimeMap.getOrDefault(topicId) {
          TopicLearningTime.newBuilder().setTopicId(topicId).build()
        }.toBuilder().apply {
          topicLearningTimeMs = if (isLastUpdatedTimestampStale(lastUpdatedTimeMs)) {
            sessionDuration
          } else topicLearningTimeMs + sessionDuration
          lastUpdatedTimeMs = oppiaClock.getCurrentTimeMs()
        }.build()
        putAggregateTopicLearningTime(topicId, topicLearningTime)
      }.build() to TopicLearningTimeActionStatus.SUCCESS
```

Bunch of small idiomatic simplifications.

Reply:
I accepted this suggestion with some minor changes.

Reply:
The changes LGTM.